packages/devtools_app/lib/src/network/http_request_data.dart

```diff
@@ -0,0 +1,246 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'http.dart';
+
+/// Contains all state relevant to completed and in-progress HTTP requests.
+class HttpRequests {
+  HttpRequests({this.requests, this.outstanding}) {
+    requests ??= [];
+    outstanding ??= {};
+  }
+
+  void clear() {
+    requests.clear();
+    outstanding.clear();
+  }
+
+  List<HttpRequestData> requests;
+
+  /// A mapping of timeline IDs to instances of HttpRequestData which are
+  /// currently in-progress.
+  Map<String, HttpRequestData> outstanding;
+}
+
+/// Used to represent an instant event emitted during an HTTP request.
+class HttpInstantEvent {
+  HttpInstantEvent._(this._rawEvent) : name = _rawEvent['name'];
+
+  final Map _rawEvent;
+  final String name;
+
+  /// The amount of time since the last instant event completed.
+  double get timeDiffMs => _timeDiffMs;
+  
+  // This is set from within HttpRequestData.
+  double _timeDiffMs;
+}
+
+/// An abstraction of an HTTP request made through dart:io.
+class HttpRequestData {
+  HttpRequestData._(this._timelineMicrosBase, this._startEvent, this._endEvent);
+
+  /// Build an instance from timeline events.
+  /// 
+  /// `timelineMicrosBase` is the offset used to determine the wall-time of a
+  /// timeline event. `events` is a list of Chrome trace format timeline
+  /// events.
+  factory HttpRequestData.fromTimeline(
+      int timelineMicrosBase, List<Map> events) {
+    Map startEvent;
+    Map endEvent;
+    final instantEvents = <Map>[];
+
+    for (final event in events) {
+      if (_isStartEvent(event)) {
+        assert(startEvent == null);
+        startEvent = event;
+      } else if (_isEndEvent(event)) {
+        assert(endEvent == null);
+        endEvent = event;
+      } else if (_isInstantEvent(event)) {
+        instantEvents.add(event);
+      } else {
+        assert(false, 'Unexpected event type');
+      }
+    }
+    final data = HttpRequestData._(timelineMicrosBase, startEvent, endEvent);
+    data._addInstantEvents(
+        [for (final instant in instantEvents) HttpInstantEvent._(instant)]);
+    return data;
+  }
+
+  /// Merges the information from another [HttpRequestData] into this instance.
+  void merge(HttpRequestData data) {
+    if (data.instantEvents.isNotEmpty) {
+      _addInstantEvents(data.instantEvents);
+    }
+    if (data._endEvent != null) {
+      _endEvent = data._endEvent;
+    }
+  }
+
+  // Timeline event helpers.
+  static bool _isStartEvent(Map event) => event['ph'] == 'b';
+  static bool _isEndEvent(Map event) => event['ph'] == 'e';
+  static bool _isInstantEvent(Map event) => event['ph'] == 'n';
```

Comment:
the phase constants are in the TraceEvent class in timeline_model.dart. We should pull the TraceEvent out into a common file for both the Timeline and Networking page to share

Reply:
Thanks for pointing this out, I didn't even think to look at the timeline_model code. I've gone ahead and updated this code to use `TraceEvent`. I saw there was a comment about pulling lots of the timeline_model code out into `package:vm_service`. Maybe we should look into that before we decide to start moving things around in DevTools?